### ISSUE TITLE
admission-controller: allow imagePullSecrets in PodDefaults

### ISSUE THREAD

**Issue by yanniszark** [kubeflow/kubeflow#4315]
/kind feature

**Why you need this feature:**
We need this feature to satisfy the private registry use-case.
In the general use-case where you want to access a private registry from a K8s cluster, you have the following options:
1. Configure Docker credentials on each node. This isn't ideal as it makes images available to all users in all namespaces and it's also Docker-specific.
2. Use platform-specific registries (eg GKE with GCR). This isn't ideal as it is not portable.
3. Use imagePullSecrets. This is the recommended way (in the k8s docs) for the general case.

In addition, this will enable users to pull images from their own registries, using their own imagePullSecrets which reside in their namespace/Profile.

**Describe the solution you'd like:**
Extend PodDefaults to include ImagePullSecrets.

/cc @zabbasi  @lluunn @kkasravi @jlewi @lluunn @kunmingg

**Comment by jlewi** [comment]
Can't service accounts be associated with imagePull secrets?
https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#add-imagepullsecrets-to-a-service-account

And
http://docs.heptio.com/content/private-registries/pr-gcr.html

Is there an advantage to using an admission hook to inject image pull secrets as opposed to just attaching the secrets to the service account and then setting the service account on the pods?

**Comment by yanniszark** [comment]
@jlewi correct, they can.
However, there isn't a good way to do this in a declarative way. 
1. We would need to make changes to all applications in manifests (edit ServiceAccounts to include imagepullsecrets, maybe in a separate overlay). The big amount of ServiceAccounts makes this even more difficult.
2. In addition, using PodDefaults is IMO a better UX for users vs making them edit ServiceAccounts.

PodDefaults provide a declarative and contained way to provide imagePullSecret injection policy in one place.

**Comment by jlewi** [comment]
TL;DR I think this raises 3 key questions

1. Is it better to attach image pull secrets to pods or service accounts?
1. is it better to inject secrets when building the spec/YAML file or using an admission hook?
1. What is the priority and how quickly do we need a short term solution?

@yanniszark Leaving aside the specifics of the use case for a second (i.e. Kubeflow private registry)

Attaching the credentials to the service account seems preferable to using an admission hook

* It seems like the better pattern; i.e. to reason about what credentials does a pod have the flow should be "what service account does the pod use? What permissions does that service account?"

  * Using a variety of mechanisms to inject credentials in a different way makes it harder to reason about.

* In general PodDefaults and mutating admission controllers seems like a pattern that could become easy to abuse; so I think we should try to avoid becoming overly reliant on it

  * i.e. the PodDefaults might be declarative; but from the perspective of an individual application/pod; the spec one specifies starts to diverge from the one that actually gets deployed; this increases as we start injecting more stuff

  * If you have all of these different admission hooks injecting things into the pod it can become difficult to reason about.

Given the above

* Is there a way we can make attaching the secrets to the service accounts easier?

   * Should we define/enforce a requirement that all Kubeflow applications need to define a parameter for image pull secrets that need to be attached to any service accounts they define?

* One alternative would be to have KFDef & KFCtl do this or kustomize

   * whether we do this in KfCtl or Kustomize the pattern would basically be create a custom plugin/generator for customize (cc @kkasravi) that would eventually transform e.g. service account to add the specified secret

   * This is effectively what an admission hook is doing

**Comment by kkasravi** [comment]
@jlewi @yanniszark 

A ServiceAccount such as
```
apiVersion: v1
kind: ServiceAccount
metadata:
  name: kf-app-XXX
imagePullSecrets:
- name: docker
- name: gcr
secrets:
- name: github
- name: gcp
```

I haven't explored the PodDefault idea, but ServiceAccounts like the one above
aggregate multiple credentials to access docker, gcr registries, github and gke.
Multiple ServiceAccounts can be added to a Rolebinding.subjects.

At least in one case under tekton, it expects to use a secret of a given name (kaniko) in the serviceaccount  in order to build and push an image. To this extent not deviating from 
ServiceAccount's role in aggregating credentials seems prudent.

Using kustomize we can declaratively add to ServiceAccount.imagePullSecrets and 
ServiceAccount.secrets by using overlays. Each overlay would apply 
a patchJson6902 'add' operation that would append to either imagePullSecrets or secrets.

The generator suggestion could point to a config file where a KfDef.spec.applications entries parameters could include secrets required (for example). This would of been a good 
example of where having a string or array would of been utilized in the parameters eg 
parameters: {{ name: 'secrets', value: ['gcp', 'github']}, { name: 'imagePullSecrets', value: ['gcr','docker']}}

**Comment by jlewi** [comment]
@yanniszark thoughts? Are you ok with using ServiceAccounts and ImagePullSecrets?

**Comment by LeonardAukea** [comment]
@jlewi There is one evident problem with this sa apporach:  

As new users join in and create their own namespace via UI. you have to manually add the ImagePullSecrets afterwards. 

Following this pattern.
```
kubectl get secret base-dockercfg --namespace=kubeflow -o yaml --export | kubectl apply --namespace=$NAMESPACE -f -
kubectl patch sa default-editor -n $NAMESPACE -p '"imagePullSecrets": [{"name": "base-dockercfg" }]'
```
This is a complete hassle with a lot of users, which I have not found a neat solution for.

**Comment by jlewi** [comment]
@LeonardAukea couldn't you just write an agent that watches for Kubeflow namespaces and then modifies the default-editor service account to include the imagePullSecrets?

**Comment by jtfogarty** [comment]
@jlewi how will the new GCP install for private GKE cluster in a private VPC with a private repo accommodate for this?

**Comment by jlewi** [comment]
> @jlewi how will the new GCP install for private GKE cluster in a private VPC with a private repo accommodate for this?

@jtfogarty accomodate what? The private GKE install works by mirror'ing all KF images to GCR which are accessible from private GKE.

If you want to continue to use non GCR registries you would have to configure networking to support that; e.g. use Cloud Nat to make them accessible.

**Comment by jtfogarty** [comment]
@jlewi Unfortunately gcr.io is blocked.  All images are required to be pulled from a private repo.

**Comment by jlewi** [comment]
> @jlewi Unfortunately gcr.io is blocked. All images are required to be pulled from a private repo. 

What do you mean gcr.io is blocked? I assume you mean you are using VPC-SC and so only gcr.io corresponding to registries within the perimeter are accessible. In that case you would use the mirror'ing functionality that we provide to mirror all KF images to your private registry.

If you are talking about accessing non KF images that's a non goal for us. You could presumably use the mirror'ing logic to mirror those images as well but we aren't trying to support that explicitly. 

@jtfogarty Supporting  GCR doesn't seem like its connected to this issue in terms of imagePullSecrets. Please open up an issue in kubeflow/gcp-blueprints if you want to discuss further.
